### PR TITLE
feat(buffer): add Node.js compatible Buffer module

### PR DESCRIPTION
**Pull request by sgtcoolguy**
**JIRA:** https://jira.appcelerator.org/browse/TIMOB-26573

**Description:**
This adds a Node-compatible `buffer` module in JS.

In Node, a `Buffer` represents a byte array (specifically an unsigned byte array). In Titanium, `Ti.Buffer` represents the exact same concept. In both cases these APIs were created because JS engines did not yet have the concept, but have long ago added TypedArrays (or more specifically in this case, a `Uint8Array`).

This shim is intended to expose Node's API, which is backed by a `Ti.Buffer` internally. It's important to note that this could result in performance issues as a result, because to support array style index setter/getters I also had to wrap instances in a JS `Proxy`. So to read/write specific bytes you end up going through a JS Proxy, then traversing our JS/native binding, grabbing the value natively, returning it across the JS/native binding layer and then back out the JS Proxy. So clearly on a per-byte level this isn't ideal. But obviously a number of operations are pure JS, and there are methods that support a little bit of "bulk" inserts to help avoid the penalties.

I think long term we should investigate possibly doing a few things:
- When not explicitly wrapping an existing `Ti.Buffer`, use a faster implementation that is backed by `Uint8Array` a la Browserify: https://github.com/feross/buffer/blob/master/index.js
- Support accepting `Uint8Array` for any Titanium APIs that take in a `Ti.Buffer`
- Possibly deprecate `Ti.Buffer` afterwards

**Caveats**
This does not include:
- `.buffer` property
- `#lastIndexOf()`
- `#read/writeBigInt` method variants (we do not yet support `BigInt`)
- `#transcode()`
- `.INSPECT_MAX_BYTES`
- `.kMaxLength`
- `.constants.MAX_LENGTH`
- `.constants.MAX_STRING_LENGTH`

The last 5 methods/properties are special in that they're exposed by the module returned by `require('buffer')` but not buffer instances or the global `Buffer` object.

I'm sure there are also other edge cases creating a Buffer from an ArrayBuffer/SharedArrayBuffer/Uint8Array that do not yet work.

### CODE REVIEW

File: tests/Resources/buffer.addontest.js [hunk 1]
```diff
@@ -0,0 +1,1647 @@
...
+		});
+
+		it('throws when trying to write offset out of range', () => {
+			const buf = Buffer.from([ 1, -2 ]);
+			should.throws(() => {
+				buf.writeUIntBE(0x1234567890ab, 1, 6);
+			}, RangeError);
+		});
+
+		// it('throws when trying to write values out of range', () => {
```
Comment (janvennemann) [inline tests/Resources/buffer.addontest.js]: Same here, can you add a fixme note and explain why this is commented out?

**Review by janvennemann — CHANGES_REQUESTED** [review]
Did a first pass and this looks really solid! Just a few minor notes.

I'll continue to test this in my WebSocket/socket.io module.

File: common/Resources/ti.internal/extensions/node/buffer.js [hunk 2]
```diff
@@ -0,0 +1,1559 @@
...
+		this._isBuffer = true;
+		// FIXME: Support .buffer property that holds an ArrayBuffer!
+	}
+
+	/**
+	 * 0 is returned if target is the same as buf
+	 *  is returned if target should come before buf when sorted.
+	 * -1 is returned if target should come after buf when sorted.
+	 * @param {Buffer} target Buffer to compare against
+	 * @param {integer} [targetStart=0] index to start in target
```
[the remark this answers is not in the document]
Comment (sgtcoolguy) [reply common/Resources/ti.internal/extensions/node/buffer.js]: Technically JSDoc can support more specialized numeric types; https://www.qooxdoo.org/current/pages/development/api_jsdoc_ref.html#handling-of-data-types

Basically I specify integer when possible because we really do expect an integer here and not a float/double - it gives more useful context/info.

File: common/Resources/ti.internal/extensions/node/buffer.js [hunk 3]
```diff
@@ -0,0 +1,1559 @@
...
+	// buf.readBigInt64LE([offset])
+	// buf.readBigUInt64BE([offset])
+	// buf.readBigUInt64LE([offset])
+	// buf.writeBigInt64BE(value[, offset])
+	// buf.writeBigInt64LE(value[, offset])
+	// buf.writeBigUInt64BE(value[, offset])
+	// buf.writeBigUInt64LE(value[, offset])
+}
+
+Buffer.allocUnsafe = (length) => {
```
[the remark this answers is not in the document]
Comment (sgtcoolguy) [reply common/Resources/ti.internal/extensions/node/buffer.js]: Because I totally forgot about static method support :)

File: common/Resources/ti.internal/extensions/node/buffer.js [hunk 4]
```diff
@@ -0,0 +1,1559 @@
...
+			if (length > 1 && string.charAt(length - 1) === '=') {
+				length--;
+			}
+			return Math.floor((length * 3) / 4); // drop fractional value
+	}
+	return utf8ByteLength(string);
+};
+
+Buffer.compare = (buf1, buf2) => {
+	// TODO: Throw a TypeError if they're not buffers? Wrap UInt8Array args in buffers?
```
[the remark this answers is not in the document]
Comment (sgtcoolguy) [reply common/Resources/ti.internal/extensions/node/buffer.js]: not quite, because I don't think we can do simple `instanceof` checks here. I think I need to do `Buffer.isBuffer()`

File: common/Resources/ti.internal/extensions/node/buffer.js [hunk 5]
```diff
@@ -0,0 +1,1559 @@
...
+			}
+
+			value.copy(buffer, 0, 0, length);
+			return buffer;
+		}
+		if (value.apiName && value.apiName === 'Ti.Buffer') {
+			return newBuffer(value);
+		}
+	}
+	throw new TypeError('The \'value\' argument must be of type: \'string\', \'Array\', \'Buffer\'');
```
[the remark this answers is not in the document]
Comment (sgtcoolguy) [reply common/Resources/ti.internal/extensions/node/buffer.js]: That method checks using typeof versus a single type. So I can't use it here...

File: common/Resources/ti.internal/extensions/node/buffer.js [hunk 6]
```diff
@@ -0,0 +1,1559 @@
...
+		if (encoding === 'base64') {
+			const blob = Ti.Utils.base64decode(value);
+			const blobStream = Ti.Stream.createStream({ source: blob, mode: Ti.Stream.MODE_READ });
+			const buffer = Ti.Stream.readAll(blobStream);
+			blobStream.close();
+			return newBuffer(buffer);
+		} else if (encoding === 'hex') {
+			return Buffer.from(stringToHexBytes(value));
+		}
+		const type = getTiCodecCharset(encoding);
```
[the remark this answers is not in the document]
Comment (sgtcoolguy) [reply common/Resources/ti.internal/extensions/node/buffer.js]: 👍

File: tests/Resources/buffer.addontest.js [hunk 7]
```diff
@@ -0,0 +1,1647 @@
...
+		});
+
+		it('handles simple ascii character fill', () => {
+			const b = Buffer.allocUnsafe(50).fill('h');
+			should(b.toString()).eql('hhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhh');
+		});
+
+		it('handles 2-byte utf-8 character fill', () => {
+			const b = Buffer.allocUnsafe(3).fill('\u0222');
+			should(b[0]).eql(0xc8); // FIXME: get expected undefined to equal 200
```
[the remark this answers is not in the document]
Comment (sgtcoolguy) [reply tests/Resources/buffer.addontest.js]: Outdated FIXME. I removed this and the others in this file

File: tests/Resources/buffer.addontest.js (hunk 1 above)
Comment (sgtcoolguy) [reply tests/Resources/buffer.addontest.js]: I'll add it back and get it right. I got a bit fatigued of writing all these tests!

File: tests/Resources/buffer.addontest.js [hunk 8]
```diff
@@ -0,0 +1,1647 @@
...
+			should(buf.readUInt16BE).be.a.Function;
+		});
+
+		it('returns unsigned value', () => {
+			const buf = Buffer.from([ 0x12, 0x34, 0x56 ]);
+			should(buf.readUInt16BE(0).toString(16)).eql('1234');
+			should(buf.readUInt16BE(1).toString(16)).eql('3456');
+		});
+
+		// it('returns signed value coerced to unsigned', () => {
```
[the remark this answers is not in the document]
Comment (sgtcoolguy) [reply tests/Resources/buffer.addontest.js]: I got lazy and didn't want to write the test properly :)

**Comment by sgtcoolguy** [comment]
@janvennemann Ok, I pushed changes to address the review comments.

File: common/Resources/ti.internal/extensions/node/buffer.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (janvennemann) [reply common/Resources/ti.internal/extensions/node/buffer.js]: 👍

File: common/Resources/ti.internal/extensions/node/buffer.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (janvennemann) [reply common/Resources/ti.internal/extensions/node/buffer.js]: Ah, damn! For some reason i thought that function includes more type checks than it actually does.

File: common/Resources/ti.internal/extensions/node/buffer.js [hunk 9]
```diff
@@ -0,0 +1,1575 @@
...
+	 * @returns {boolean}
+	 */
+	static isBuffer(obj) {
+		return obj !== null && obj !== undefined && obj._isBuffer === true;
+	}
+}
+
+Buffer.poolSize = 8192;
+
+export default Buffer;
```
Comment (janvennemann) [inline common/Resources/ti.internal/extensions/node/buffer.js]: The export is not compatible with the node `buffer` module this way. The core module exports it as `Buffer` instead of default.

Some third-party node modules do the following for example:

```js
var Buffer = require('buffer').Buffer;
```

**Review by janvennemann — DISMISSED** [review]

File: common/Resources/ti.internal/extensions/node/buffer.js (hunk 9 above)
Comment (sgtcoolguy) [reply common/Resources/ti.internal/extensions/node/buffer.js]: good catch, pushing commits to fix this now

**Comment by sgtcoolguy** [comment]
squash merged manually
